### PR TITLE
SAMZA-1952: StreamPartitionCountMonitor for standalone.

### PR DESCRIPTION
**Pull request by shanthoosh**
This patch adds the capability to detect the partition change of the input streams of a stateless standalone jobs and trigger a re-balancing phase(which will essentially account for new partitions from input stream and distribute it to the live processors of the group). 

Existing partition count detection of input streams is broken in yarn for stateful jobs. This will be addressed for both yarn and standalone as a part of #622

### CODE REVIEW

File: samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java [hunk 1]
```diff
@@ -642,4 +646,73 @@ public void testShouldStopStreamApplicationWhenShutdownTimeOutIsLessThanContaine
...
+
+    // Increase the partition count of input kafka topic to 100.
+    AdminUtils.addPartitions(zkUtils(), inputKafkaTopic, 100, "", true, RackAwareMode.Enforced$.MODULE$);
+
+    publishKafkaEvents(inputKafkaTopic, 0, NUM_KAFKA_EVENTS, PROCESSOR_IDS[0]);
+
+    CountDownLatch kafkaEventsConsumedLatch2 = new CountDownLatch(NUM_KAFKA_EVENTS);
+    CountDownLatch processedMessagesLatch2 = new CountDownLatch(1);
+
+    ApplicationRunner appRunner2 = ApplicationRunners.getApplicationRunner(TestStreamApplication.getInstance(
```
Comment (sborya) [inline samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java]: why do you need to start another processor? The monitor should trigger job model regeneration, shouldn't it?

File: samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java [hunk 2]
```diff
@@ -320,16 +326,34 @@ private JobModel generateNewJobModel(List<String> processors) {
...
       zkUtils.subscribeToProcessorChange(new ProcessorChangeHandler(zkUtils));
-      debounceTimer.scheduleAfterDebounceTime(ON_PROCESSOR_CHANGE, debounceTimeMs, () -> {
-          // actual actions to do are the same as onProcessorChange
-          doOnProcessorChange(new ArrayList<>());
-        });
+      if (!new StorageConfig(config).hasDurableStores()) {
+        streamPartitionCountMonitor = getPartitionCountMonitor();
+        streamPartitionCountMonitor.start();
+      }
+      debounceTimer.scheduleAfterDebounceTime(ON_PROCESSOR_CHANGE, debounceTimeMs, () -> doOnProcessorChange());
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java]: Sure, will make this change.

File: samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java [hunk 3]
```diff
@@ -477,6 +499,11 @@ public void handleStateChanged(Watcher.Event.KeeperState state) {
           if (leaderElector.amILeader()) {
             leaderElector.resignLeadership();
           }
+
+          if (streamPartitionCountMonitor != null) {
+            streamPartitionCountMonitor.stop();
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java]: Not sure if that's necessary. `StreamPartitionCountMonitor.stop()` is idempotent.

File: samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java (hunk 1 above)
Comment (shanthoosh) [reply samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java]: Will change it. Thanks for pointing it out.

File: samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java [hunk 4]
```diff
@@ -642,4 +648,59 @@ public void testShouldStopStreamApplicationWhenShutdownTimeOutIsLessThanContaine
...
+
+    // Validate that the input partition count is 5 in the JobModel.
+    Assert.assertEquals(5, ssps.size());
+
+    // Increase the partition count of input kafka topic to 100.
+    AdminUtils.addPartitions(zkUtils(), inputKafkaTopic, 100, "", true, RackAwareMode.Enforced$.MODULE$);
+
+    while (Objects.equals(zkUtils.getJobModelVersion(), jobModelVersion)) {
+      LOGGER.info("Waiting for new jobModel to be published");
+      Thread.sleep(1000);
```
Comment (sborya) [inline samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java]: to make it little bit faster, may be you can make the sleep time progressive (10,20,40,80..)?

File: samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java]: Still, it is cleaner. Stop() doesn't have to be idempotent and may get changed in the future.

[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java]: Done.

File: samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java (hunk 4 above)
Comment (shanthoosh) [reply samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java]: Sounds like a good idea. Added the exponential backoff based sleep.

**Pull request closed without merging**